src/main/java/com/glencoesoftware/omero/ms/image/region/ImageRegionVerticle.java

```diff
@@ -138,45 +292,311 @@ private void renderImageRegion(Message<String> message) {
+                        if (result2.succeeded()) {
+                            future.complete(result2.result());
+                        } else {
+                            future.fail(result2.cause());
+                        }
+                    });
+                }
             } else {
-                message.reply(imageRegion);
+                future.fail(result1.cause());
             }
-        } catch (PermissionDeniedException
-                | CannotCreateSessionException e) {
-            String v = "Permission denied";
-            log.debug(v);
-            message.fail(403, v);
-        } catch (IllegalArgumentException e) {
-            log.debug(
-                "Illegal argument received while retrieving image region", e);
-            message.fail(400, e.getMessage());
-        } catch (Exception e) {
-            String v = "Exception while retrieving image region";
-            log.error(v, e);
-            message.fail(500, v);
+        });
+
+        timer.observeDuration();
+        return future;
+    }
+
+    /**
+     * Render image region.
+     * @param imageRegionCtx request context
+     * @param request OMERO request based on the current context
+     * @return Future which will be completed with the image region byte array.
+     */
+    private Future<byte[]> renderImageRegion(
+            ImageRegionCtx imageRegionCtx, Supplier<OmeroRequest> request) {
+        Summary.Timer timer = renderImageRegionSummary.startTimer();
+        Future<byte[]> future = Future.future();
+
+        ImageRegionRequestHandler requestHandler =
+                new ImageRegionRequestHandler(
+                        imageRegionCtx, context, families,
+                        renderingModels, lutProvider);
+        Future<Pixels> step1 =
+                getPixels(imageRegionCtx, request, requestHandler);
+
+        step1.compose(pixels -> {
+            try {
+                if (pixels == null) {
+                    throw new ObjectNotFoundException(
+                            "Cannot find Image:" + imageRegionCtx.imageId);
+                }
+                requestHandler.setPixels(pixels);
+                byte[] imageRegion = request.get().execute(
+                        requestHandler::renderImageRegion);
+                if (imageRegion == null) {
+                    throw new ObjectNotFoundException("Cannot render region");
+                }
+                // Cache the image region
+                JsonObject setMessage = new JsonObject();
+                setMessage.put("key", imageRegionCtx.cacheKey());
+                setMessage.put("value", imageRegion);
+                vertx.eventBus().send(
+                        RedisCacheVerticle.REDIS_CACHE_SET_EVENT,
+                        setMessage);
+                future.complete(imageRegion);
+            } catch (Exception e) {
+                future.fail(e);
+            }
+        }, future);
+
+        timer.observeDuration();
```

Comment:
With all the async code above, observing the duration here won't tell us anything, it will end up being 0ms.